snuba/utils/streams/producer.py

```diff
@@ -0,0 +1,30 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from concurrent.futures import Future
+from typing import Generic, NamedTuple, Union
+
+from snuba.utils.streams.types import Partition, TPayload, Topic
+
+
+class MessageDetails(NamedTuple):
+    partition: Partition
+    offset: int
+
+
+class Producer(Generic[TPayload], ABC):
+    @abstractmethod
+    def produce(
+        self, destination: Union[Topic, Partition], payload: TPayload
+    ) -> Future[MessageDetails]:
```

Comment:
This was the cleanest API that I could think of, but I am open to alternatives. I considered trying to use `typing.overload` but I'm not familiar with it and this seemed straightforward enough to me.